### DISCUSSION TITLE
Problem using @filename

### DISCUSSION THREAD

**Discussion by jamesthenry** [qpdf/qpdf#688] [Q&A] [answered]
Greetings,

I am trying to concatenate some pdf files.  It works if I put all options on the command line, like so:

qpdf --empty --pages "a.pdf" "b.pdf" "c.pdf" "d.pdf" -- combined.pdf

But I have problems if I try using the @filename feature.

My  command line is:

qpdf @qopts2.txt

The contents of qopts2.txt are:

(first line follows)
--empty
--pages
"a.pdf"
"b.pdf"
"c.pdf"
"d.pdf"
\--
combined2.pdf
(final line precedes)

The result I get is:  qpdf: error at * in numeric range *"b.pdf": unexpected character

There is no whitespace in any of the lines in qopts2.txt

Can anyone tell me what I'm doing wrong and how to fix it?

Thanks!

**Comment by jberkenbilt** [comment, marked as answer]
It's because it's interpreting the `"` characters as a literal part of the file name, and since there is no file with the name `"a.pdf"` (etc.), it is trying to treat that as a page range. Just remove the `"` characters from your file.

**Reply by jamesthenry** [reply]
That's it - thanks!  Thought double quotes might be needed in the event of filenames with embedded blanks, but that case is handled correctly.

Thanks for the response.

**Reply by defencedog** [reply]
`@filename` procedure deletes all bookmarks?

**Comment by jberkenbilt** [comment]
It doesn't have anything to do with `@filename`. It has to do with qpdf's present lack of preservation of bookmarks from all but the first file specified, which in this case is `--empty`. This is a long-standing issue and the most requested feature of qpdf, but it is very hard to fix. For an in-depth discussion, see #1104. The present reality is that you can replace `--empty` with any of the files listed in `--pages` while leaving the arguments to `--pages` as they are. If you do that, the bookmarks will be preserved from _that file only_. There is presently no way to get qpdf to preserve bookmarks from all files. I hope that will eventually work, but it depends on me or someone else having a substantial chunk of time to work through the design of the feature and to implement it. It doesn't seem like it should be complicated, but it is because of the way bookmarks are stored in PDF files.

**Reply by defencedog** [reply]
I used `cpdf` binary on aarch64 instead ...its freemium but it works & provides capability to 
- preserve individual file bookamarks
- create bookmarks from filenames of concatenated pdf docs

> https://www.coherentpdf.com/cpdfmanual272.pdf Chapter#6

**Comment by jberkenbilt** [comment]
Yeah, cpdf can do some things qpdf can't do. It's a commercial tool with people who are paid to maintain it.

BTW, I'm a major π fan and like your avatar. :-)
